dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/util/SerializerWithLimits.java

```
@@ -125,24 +151,73 @@ public void serialize(Object value, String type, Limits limits) throws Exception
       } else {
         isComplete = serializeObjectArray((Object[]) value, limits);
       }
-      typeSerializer.arrayEpilogue(value, isComplete, arraySize);
+      tokenWriter.arrayEpilogue(value, isComplete, arraySize);
     } else if (value instanceof Collection && (limits.maxReferenceDepth > 0)) {
-      typeSerializer.collectionPrologue(value);
+      tokenWriter.collectionPrologue(value);
       Collection<?> col = (Collection<?>) value;
       boolean isComplete = serializeCollection(col, limits);
-      typeSerializer.collectionEpilogue(value, isComplete, col.size());
+      tokenWriter.collectionEpilogue(value, isComplete, col.size());
     } else if (value instanceof Map && (limits.maxReferenceDepth > 0)) {
-      typeSerializer.mapPrologue(value);
+      tokenWriter.mapPrologue(value);
       Map<?, ?> map = (Map<?, ?>) value;
       Set<? extends Map.Entry<?, ?>> entries = map.entrySet();
       boolean isComplete = serializeMap(entries, limits);
-      typeSerializer.mapEpilogue(map, isComplete);
+      tokenWriter.mapEpilogue(map, isComplete);
     } else if (limits.maxReferenceDepth > 0) {
-      typeSerializer.objectValue(value, this, limits);
+      serializeObjectValue(value, limits);
     } else {
-      typeSerializer.reachedMaxDepth();
+      tokenWriter.reachedMaxDepth();
+    }
+    tokenWriter.epilogue(value);
+  }
+
+  private void serializeObjectValue(Object value, Limits limits) throws Exception {
+    tokenWriter.objectPrologue(value);
```

maybe we should put a outer try catch to ensure that if write objectProlouge we also write ObjectEpilouge. 
Because we have a generic Exception e catch it could swallow errors and break structure of the json.

---

if we have an exception it means the process is broken. what is the point to finish the structure if we will not use it anyway?
having said that, I need to remove the try/catch in `onField` method